services/src/main/scala/cromwell/services/metadata/impl/MetadataDatabaseAccess.scala

```diff
@@ -178,14 +178,73 @@ trait MetadataDatabaseAccess {
       queryParameters.statuses, queryParameters.names, queryParameters.ids.map(_.toString), queryParameters.labels.map(label => (label.key, label.value)),
       queryParameters.startDate.map(_.toSystemTimestamp), queryParameters.endDate.map(_.toSystemTimestamp))
 
-    workflowSummaryCount flatMap { count =>
-      workflowSummaries map { workflows =>
-        (WorkflowQueryResponse(workflows.toSeq map { workflow =>
-          MetadataService.WorkflowQueryResult(id = workflow.workflowExecutionUuid, name = workflow.workflowName, status = workflow.workflowStatus, start = workflow.startTimestamp map { _.toSystemOffsetDateTime }, end = workflow.endTimestamp map { _.toSystemOffsetDateTime })
-        }),
-          //only return metadata if page is defined
-          queryParameters.page map { _ => QueryMetadata(queryParameters.page, queryParameters.pageSize, Option(count)) })
+    def queryMetadata(count: Int): Option[QueryMetadata] = {
+      queryParameters.page map { _ => QueryMetadata(queryParameters.page, queryParameters.pageSize, Option(count)) }
+    }
+
+    def summariesToQueryResults(workflows: Traversable[WorkflowMetadataSummaryEntry]): Future[Traversable[MetadataService.WorkflowQueryResult]] = {
+      val mapped: Traversable[Future[MetadataService.WorkflowQueryResult]] = workflows map { summaryToQueryResult }
+      val result: Future[Traversable[MetadataService.WorkflowQueryResult]] = Future.sequence(mapped)
+      result
+    }
+
+    def summaryToQueryResult(workflow: WorkflowMetadataSummaryEntry): Future[MetadataService.WorkflowQueryResult] = {
+
+      def queryResult(labels: Map[String, String], parentId: Option[String]): MetadataService.WorkflowQueryResult = {
+        MetadataService.WorkflowQueryResult(
+          id = workflow.workflowExecutionUuid,
+          name = workflow.workflowName,
+          status = workflow.workflowStatus,
+          start = workflow.startTimestamp map { _.toSystemOffsetDateTime },
+          end = workflow.endTimestamp map { _.toSystemOffsetDateTime },
+          labels = if (labels.nonEmpty) { Some(labels) } else None,
+          parentWorkflowId = parentId
```

Comment:
I think it'd be nicer to wrap the "do we want labels" and "do we want parentWorkflowId" into the original metadata query, so that we only need to go to the database once (in this scheme I think we might end up querying the entire metadata table *three* times per QUERY request).

An added benefit would be that these two lines would be easier to write if the fields were already part of `WorkflowQueryResult`

Reply:
Do you mean add these fields to the `WorkflowMetadataSummaryEntry` model? I agree it would be more efficient to get all of the information from the database at the same time, but if we want to support more "additional fields" that would require adding each field to the database model.

Reply:
IIRC the first query hits the summary table, but you're right that these follow on queries go against the potentially enormous metadata table.